dace/transformation/interstate/fpga_global_to_local.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
+""" Contains a transformation that changes the storage type of a Global FPGA variable to Local when certains
+    conditions are met.
+"""
+
+import networkx as nx
+
+from dace import registry, properties
+from dace.transformation import transformation
+from dace import symbolic, dtypes
+from dace.sdfg import nodes, trace_nested_access
+from dace import config
+
+
+@registry.autoregister
+@properties.make_properties
+class FPGAGlobalToLocal(transformation.Transformation):
+    """ Implements the FPGAGlobalToLoca transformation, which takes an entire
+        SDFG  and changes the storage type of a Global FPGA data container to Local in the following situation:
+        - data the is transient,
+        - and the data is not a transient shared with other states,
+        - and data has a compile-time known size. """
+    @staticmethod
+    def annotates_memlets():
+        return True
+
+    @staticmethod
+    def expressions():
+        # Match anything
+        return [nx.DiGraph()]
+
+    @staticmethod
+    def can_be_applied(graph, candidate, expr_index, sdfg, strict=False):
+        # It can always be applied
+        return True
+
+    @staticmethod
+    def match_to_str(graph, candidate):
+        return graph.label
+
+    def apply(self, sdfg):
+
+        count = 0
+
+        for sd, name, desc in sdfg.arrays_recursive():
+            if desc.transient and name not in sd.shared_transients(
+            ) and desc.storage == dtypes.StorageType.FPGA_Global:
```

Comment:
Is there any way we can make this apply to only a single array? I'm not sure the matching supports this. Maybe we can use a random node as our match?

Reply:
Why do you want to do this? To apply it repeatedly?

Reply:
So you don't have to promote everything if you don't want to. For example, one array could be constant-size, but **HUGE**, so you don't want to put it in local memory

Reply:
Makes sense. Not sure how it should like at the end: the matching returns the list of the potential matches and then it's up to the user to decide to which container apply the transformation?

Reply:
No, I was thinking of one match per container that can be converted :-)

Reply:
So that it can be used with apply_repeatedly?

Reply:
Yes if desired :-) Or someone can get all matches, and search for the one with the right array name in it programmatically

Reply:
I'm honestly struggling in understanding how to do this (get all the matches). Can you please point me to any transformation that does this so I can take some inspiration? :grin:

Reply:
Maybe [local storage](https://github.com/spcl/dace/blob/master/dace/transformation/dataflow/local_storage.py) is relevant?